ivan_selivanov/SUBS.scala

```diff
@@ -0,0 +1,6 @@
+import scala.io.StdIn.readLine
+
+object SUBS extends App {
+  val s = readLine()
```

Comment:
Do we really need to use regular expressions with lookahead here? Why not just a substring search?

Reply:
Well, in my opinion regex is perfect for those kind of tasks, or am I wrong? Please explain the benefit of implementing substring search. I understand when regex is complicated and becoming not readable, but this regex is fairly short and simple.

Reply:
Regex is a language, so it adds extra complexity to the code (depends on the reader).   
I usually work with Perl regular expressions, so I am not familiar with this syntax.
The documentation says it's just `X, via zero-width positive lookahead`.  I don't sure that this is simple for everyone.

> regex is fairly short and simple

`Groups` is an advanced part of regexp.  

Usually, for that kind of matching, we don't need groups.  Could you please, share an idea, why do we need groups here? 

The second thing about regexp is performance. `Substring search` sounds easier than `nondeterministic finite automaton`.  And it works much faster.  In the context of that task, performance does not matter but in the context of real nucleotide sequences, it does.

Reply:
First of all, thank you for your detailed response. I used regex for pattern mathing, and since it pretty much was created for it, I thought it will the best option. But you are right, it works slower than substring search, difference it's not that big 16.97 vs 18.48 sec on a 400MB file, but still.

Reply:
1. Aside from any performance considerations, the problem does not _need_ the use of regex. Regex is naturally used when there are classes of strings to be used. And when you read regex solution, you expect that this kind of problem is solved here. However, this is just a substring search, so it's natural and more clear to understand, if you used a substring search.

2. Regex requires building a state machine at the beginning, which requires additional time and memory.

3. Complexity of substring search is O(N). Regex "findAll" is at least O(N), can be worse depending on the pattern.

Reply:
Sure, maybe the regex is the overkill for this task. If it didn't required overlapped substrings to be accounted, than state machine would be perfect. State machine would just handle chars one by one. Substring search on the other hand require extracting substring fixed size and comparing it to the searched one. O(2N) for substring search vs O(N) for state machine. Of course state machine requires additional time and memory, but if we are talking about large sequence it should not be taken into account. 
I am not sure if it's possible to use regex efficiently in parallel computing, so if we are talking about real nucleotide sequences, substring parallel search is definitely better.

Reply:
> Substring search on the other hand require extracting substring fixed size and comparing it to the searched one
It doesn't have to. In your code, you do substring search, but you can as well use `String.indexOf(String, int)` to avoid copying.

>  I am not sure if it's possible to use regex efficiently in parallel computing, so if we are talking about real nucleotide sequences, substring parallel search is definitely better.
For large strings there are faster substring search algorithms that use lookup table to skip some characters. I don't know though what is really used for genotypes, maybe there are more specialized algorithms.

> I am not sure if it's possible to use regex efficiently in parallel computing
I'm not sure if it has any real benefit in this case (parallelize search within _one_ string). It will be typically constrained by the speed of reading this string from disk/network rather then CPU.